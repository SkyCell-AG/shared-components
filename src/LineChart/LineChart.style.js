import {
    makeStyles,
} from '@mui/styles'

const useStyles = makeStyles(() => {
    return {
        chart: {
            height: '80%',
        },
        rangeFilter: {
            height: '20%',
        },
    }
})

export default useStyles
