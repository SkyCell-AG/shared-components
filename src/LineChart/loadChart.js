/* global google */
import drop from 'lodash/drop'
import omit from 'lodash/omit'
import moment from 'moment'

import loadScript from 'utils/loadScript'

const url = 'https://www.gstatic.com/charts/loader.js'

const createColumns = (chart, chartColumns) => {
    chartColumns.forEach((column) => {
        chart.addColumn(...column)
    })
}

const fillerNulls = (columns) => {
    return columns.slice(1, columns.length - 1).map(() => {
        return null
    })
}

const loadChart = (chartData, elm, columns, options, isDateRange, onError) => {
    loadScript(url).then(() => {
        google.charts.load('current', {
            packages: [
                'corechart',
                'line',
                'controls',
            ],
        })

        const updatedChartData = chartData.map((element) => {
            return [
                new Date(element[0]),
                ...drop(element, 1),
            ]
        })

        const yesterday = moment().subtract(1, 'days').format()

        const lastDateEntry = moment(chartData[chartData.length - 1][0])
        const isAfterLastEntry = moment().isAfter(lastDateEntry)
        const start = isAfterLastEntry ? moment(lastDateEntry).subtract(1, 'days').format() : yesterday
        const end = isAfterLastEntry
            ? moment(chartData[chartData.length - 1][0]).format()
            : moment().format()

        const drawChart = () => {
            const dashboard = new google.visualization.Dashboard(elm)
            const getSeries = () => {
                return options.series.map((
                    option,
                ) => {
                    return omit(option, [
                        'pointSize',
                        'lineWidth',
                    ])
                })
            }

            const control = new google.visualization.ControlWrapper({
                controlType: 'ChartRangeFilter',
                containerId: 'rangeFilter',
                options: {
                    filterColumnIndex: 0,
                    ui: {
                        chartType: 'LineChart',
                        chartOptions: {
                            ...options,
                            chartArea: omit(options.chartArea, ['top']),
                            series: getSeries(),
                        },
                    },
                },
                state: {
                    range: {
                        start: new Date(start),
                        end: new Date(end),
                    },
                },
            })

            const chart = new google.visualization.ChartWrapper({
                chartType: 'LineChart',
                containerId: 'chart',
                options: {
                    ...options,
                },
            })

            const data = new google.visualization.DataTable()

            createColumns(data, columns)

            const date = new Date()
            const offset = date.getTimezoneOffset()

            const chartDataWithFillers = [
                ...updatedChartData,
                [
                    new Date(date.getTime() + offset * 60000),
                    0,
                    ...fillerNulls(columns),
                ],
            ]

            const updatedChartDataWithFillers = !isDateRange
                ? chartDataWithFillers : updatedChartData

            data.addRows(updatedChartDataWithFillers)

            dashboard.bind(control, chart)
            dashboard.draw(data)
        }

        return google.charts.setOnLoadCallback(drawChart)
    })
        .catch(onError)
}

export default loadChart
