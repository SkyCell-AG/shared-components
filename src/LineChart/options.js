const options = {
    animation: {
        startup: false,
        duration: 200,
    },
    focusTarget: 'category',
    legend: {
        position: 'bottom',
    },
    interpolateNulls: true,
    selectionMode: 'single',
    explorer: {
        actions: [
            'dragToZoom',
            'rightClickToReset',
        ],
        maxZoomIn: 0.001,
        maxZoomOut: 1000,
        axis: 'horizontal',
    },
    height: '100%',
    width: '100%',
    crosshair: {
        trigger: 'both',
        opacity: 0.3,
    },
    fontName: 'Roboto',
    fontSize: 13,
}

export default options
