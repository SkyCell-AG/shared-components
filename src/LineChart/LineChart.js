import React, {
    useLayoutEffect,
    useRef,
} from 'react'
import PropTypes from 'prop-types'

import loadChart from './loadChart'
import defaultOptions from './options'

const propTypes = {
    value: PropTypes.arrayOf(PropTypes.array),
    columns: PropTypes.arrayOf(PropTypes.array).isRequired,
    userOptions: PropTypes.objectOf(PropTypes.any),
    onError: PropTypes.func.isRequired,
    className: PropTypes.string,
    isDateRange: PropTypes.bool.isRequired,
}

const defaultProps = {
    value: [],
    userOptions: {},
    className: '',
}

const LineChart = (props) => {
    const {
        value,
        columns,
        onError,
        userOptions,
        className,
        isDateRange,
    } = props

    const elm = useRef()

    useLayoutEffect(() => {
        loadChart(value, elm.current, columns, {
            ...defaultOptions,
            ...userOptions,
        }, onError, isDateRange)
    }, [
        columns,
        onError,
        userOptions,
        value,
        isDateRange,
    ])

    return (
        <div
            className={className}
            ref={elm}
        />
    )
}

LineChart.propTypes = propTypes
LineChart.defaultProps = defaultProps

export default LineChart
