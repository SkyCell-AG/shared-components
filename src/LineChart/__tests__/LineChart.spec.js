import React from 'react'
import {
    shallow,
} from 'enzyme'

import LineChart from '../LineChart'

describe('app/Apps/shared-components/LineChart', () => {
    it('LineChart page snapshot', () => {
        const props = {
            columns: [[
                'number',
                'Temperature Range Minimum',
            ]],
            onError: jest.fn(),
            isDateRange: true,
        }

        const wrapper = shallow(
            <LineChart {...props} />,
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('Test', () => {
        const props = {
            columns: [[
                'number',
                'Temperature Range Minimum',
            ]],
            onError: jest.fn(),
            isDateRange: false,
        }

        const wrapper = shallow(
            <LineChart {...props} />,
        )

        expect(wrapper).toMatchSnapshot()
    })
})
