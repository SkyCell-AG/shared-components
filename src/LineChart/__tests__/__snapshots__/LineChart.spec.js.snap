// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/Apps/shared-components/LineChart LineChart page snapshot 1`] = `
<div
  className=""
>
  <div
    className="makeStyles-chart-1"
    id="chart"
  />
  <div
    className="makeStyles-rangeFilter-2"
    id="rangeFilter"
  />
</div>
`;
